examples/2-multi-step.ipynb

```diff
@@ -0,0 +1,411 @@
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# TODO: Get rid of this\n",
+    "!rm -r ./cache"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As before, we'll import a dataset from Scikit-learn for our experiments and divide it into training and test sets"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.datasets import load_breast_cancer\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "\n",
+    "dset = load_breast_cancer()\n",
+    "X_dset, y_dset = dset.data, dset.target\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X_dset, y_dset, test_size=0.2, stratify=y_dset, random_state=42)\n",
+    "\n",
+    "print('{} samples for training'.format(len(y_train)))\n",
+    "print('{} samples for test'.format(len(y_test)))\n",
+    "\n",
+    "data_train = {'input':\n",
+    "                {\n",
+    "                     'X': X_train,\n",
+    "                     'y': y_train,\n",
+    "                }\n",
+    "            }\n",
+    "\n",
+    "data_test = {'input':\n",
+    "                {\n",
+    "                     'X': X_test,\n",
+    "                     'y': y_test,\n",
+    "                }\n",
+    "            }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating pipeline components\n",
+    "\n",
+    "This time we want to make a smarter attempt at digit classification. We'll normalize our data, run PCA to compute some features of a different flavour and then combine them with our original features in a final Random Forest step."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our first step will be a normalization step. We could use the one from Scikit-learn but we'll write a pure Numpy implementation just to show how this could be done:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.externals import joblib\n",
+    "\n",
+    "class NormalizationTransformer(BaseTransformer):\n",
+    "    def __init__(self):\n",
+    "        self.mean = None\n",
+    "        self.std = None\n",
+    "    \n",
+    "    # Having only X as input ensures that we don't accidentally fit y!\n",
+    "    def fit(self, X):\n",
+    "        self.mean = np.mean(X, axis=0)\n",
+    "        self.std = np.std(X, axis=0)\n",
+    "        return self\n",
+    "\n",
+    "    def transform(self, X, **kwargs):\n",
+    "        X_tfm  = (X - self.mean) / self.std\n",
+    "        return {'X': X_tfm}\n",
+    "    \n",
+    "    def save(self, filepath):\n",
+    "        joblib.dump([self.mean, self.std], filepath)\n",
```

Comment:
@mromaniukcdl I would go with tuple instead of list since those are not elements of the same "type".